### PR TITLE
TS: A bunch of TypeScript fixes

### PR DESCRIPTION
**Pull request by asgerf**
Fixes a couple of TypeScript issues:
- As of TypeScript 3.7, type aliases are reified and extracted as `TypeReference`s, but were hard to work with in QL. It was really hard to get to the underlying type.
- Rest parameters were ignored on call signature types (at the static type level)
- We did not expose whether a parameter was declared optional (at the syntax level)

Also tsconfig compilation options were not actually passed on to the TypeScript compiler. We do now, which which means null/undefined types are extracted properly when the relevant strictness flag is set, and a few bugs surfaced which I've fixed along the way. I made the extractor explicitly not try to extract JS files using the TypeScript extractor (when `allowJs: true` is set), mainly to avoid changing too much, but it's something we might revisit in a future PR.

[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/typescript-fixes_1574111740797) looks fine, there's even a few fixed FPs.

Commit-by-commit review strongly encouraged. I'm aware that the PR does a bunch of things that could be done in separate PRs, but I'm afraid I'd be bogged down forever with conflict resolution and rebasing. The commit history is fairly straightforward.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 1]
```diff
@@ -2610,6 +2614,30 @@ class CallSignatureType extends @signature_type {
    * Gets the name of a parameter of this signature.
    */
   string getAParameterName() { result = getParameterName(_) }
+
+  /**
+   * Holds if this signature declares a rest parameter, such as `(x: number, ...y: string[])`.
+   */
+  predicate hasRestParameter() { signature_rest_parameter(this, _) } // TODO
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/TypeScript.qll]: Seems like you forgot to remove the `// TODO`, because the line looks fine.

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/TypeScript.qll]: Replace with `exists(getRestParameterArrayType())` for readability?

File: javascript/extractor/src/com/semmle/js/ast/AFunction.java [hunk 2]
```diff
@@ -29,7 +33,8 @@ public AFunction(
       List<ITypeExpression> parameterTypes,
       List<DecoratorList> parameterDecorators,
       ITypeExpression returnType,
-      ITypeExpression thisParameterType) {
+      ITypeExpression thisParameterType,
+      IntList optionalParmaeterIndices) {
```
Comment (erik-krogh) [inline javascript/extractor/src/com/semmle/js/ast/AFunction.java]: Typo. `optionalParmaeterIndices` -> `optionalParameterIndices`.

File: javascript/extractor/src/com/semmle/js/parser/TypeScriptASTConverter.java [hunk 3]
```diff
@@ -1890,6 +1896,19 @@ private ITypeExpression convertThisParameterType(JsonObject function) throws Par
...
   }
 
+  private IntList getOptionalParameterIndices(JsonObject function) throws ParseError {
+    IntList list = IntList.create(0);
+    int index = -1;
+    for (JsonElement param : getProperParameters(function)) {
+      ++index;
+      if (param.getAsJsonObject().has("questionToken")) {
+        list.add(index);
+        ;
```
Comment (erik-krogh) [inline javascript/extractor/src/com/semmle/js/parser/TypeScriptASTConverter.java]: The line contains just a `;`. 

Also, I would get rid of the `++` and `-1` in this method. 
Maybe something like (untested): 
```Java
List<JsonElement> parameters = getProperParameters(function);
for (int i = 0; i < parameters.length; i++) {
  if (param.getAsJsonObject().has("questionToken")) {
    list.add(i);
  }
}
```

I know there is overhead in creating an additional list, but its more readable.

File: javascript/ql/src/semmle/javascript/TypeScript.qll (hunk 1 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/TypeScript.qll]: I prefer to use the database relation so these classes don't have unexpected dependencies. For instance, this predicate would come to depend on the charpred of `PlainArrayType`, which itself has some non-trivial dependencies. Nothing bad would happen in this case, but that's my reason for doing it.

File: javascript/extractor/src/com/semmle/js/parser/TypeScriptASTConverter.java (hunk 3 above)
Comment (asger-semmle) [reply javascript/extractor/src/com/semmle/js/parser/TypeScriptASTConverter.java]: Oh right, the dangers of not checking the output after auto-formatting 😅 

> I know there is overhead in creating an additional list, but its more readable.

Well I guess we just see things very differently here. I don't really want `getProperParameters` to build a copy every time it is called.

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 4]
```diff
@@ -1829,6 +1829,27 @@ class Type extends @type {
     )
   }
 
+  /**
+   * Repeatedly unfolds unions, intersections, and type aliases and gets any of the underlying types,
+   * or this type itself if it is not a union or intersection.
+   *
+   * For example, the type `(S & T) | U` unfolds to `S`, `T`, and `U`.
+   *
+   * If this is a type alias, the alias is itself included in the result, but this is not the case for intermediate type aliases.
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/TypeScript.qll]: I'm a bit conflicted about including the type alias itself in the result, and not including the union/intersection type in the result.  

Do you have any particular reason for including the type alias itself in the result?

Comment (asgerf) [reply javascript/ql/src/semmle/javascript/TypeScript.qll]: One of the really important use-cases is to recognize expressions based on some named type, and the user should be able to specify the name of a type alias here. Like if you want to find all expressions of type `Electron.BrowserWindow` you shouldn't have to worry about whether `BrowserWindow` is a type alias.

So I actually wanted to include all intermediate type aliases, but currently TypeScript only retains the outermost aliased name.

I guess we could rephrase it to say that we do a "best effort" to include all intermediate type aliases. WDYT?

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/TypeScript.qll]: > Like if you want to find all expressions of type Electron.BrowserWindow you shouldn't have to worry about whether BrowserWindow is a type alias.

That is a good motivation for why type-aliases are kept, and untion/intersection types are not. :+1: 

> I guess we could rephrase it to say that we do a "best effort" to include all intermediate type aliases. WDYT?

I think I like the current phrasing more.
